File: frontend/src/api/notifications.ts
/**
 * 消息/通知 API
 */
import client from './client';

export interface Notification {
  id: number;
  type: string;
  title: string;
  message: string;
  priority: string;
  is_read: boolean;
  read_at?: string;
  related_id?: string;
  related_type?: string;
  created_at: string;
  updated_at: string;
}

/**
 * 获取消息列表
 */
export const getNotifications = async (params?: {
  skip?: number;
  limit?: number;
  is_read?: boolean;
  type?: string;
}): Promise<Notification[]> => {
  const response = await client.get('/notifications/', { params });
  return response.data;
};

/**
 * 获取未读消息数
 */
export const getUnreadCount = async (): Promise<{ unread_count: number }> => {
  const response = await client.get('/notifications/unread-count');
  return response.data;
};

/**
 * 获取单个消息
 */
export const getNotification = async (notificationId: number): Promise<Notification> => {
  const response = await client.get(`/notifications/${notificationId}`);
  return response.data;
};

/**
 * 标记消息为已读
 */
export const markAsRead = async (notificationId: number): Promise<Notification> => {
  const response = await client.put(`/notifications/${notificationId}/read`);
  return response.data;
};

/**
 * 标记所有消息为已读
 */
export const markAllAsRead = async (): Promise<{ status: string }> => {
  const response = await client.put('/notifications/mark-all-as-read');
  return response.data;
};

/**
 * 删除消息
 */
export const deleteNotification = async (notificationId: number): Promise<{ status: string }> => {
  const response = await client.delete(`/notifications/${notificationId}`);
  return response.data;
};

/**
 * 删除所有消息
 */
export const deleteAllNotifications = async (): Promise<{ status: string }> => {
  const response = await client.delete('/notifications/');
  return response.data;
};

File: backend/app/api/v1/notifications.py
"""
消息/通知 API 端点
"""
from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session
from typing import List, Optional
from datetime import datetime
from pydantic import BaseModel

from app.database import get_db
from app.models.notification import Notification, NotificationTypeEnum, NotificationPriorityEnum
from app.models.user import User
from app.dependencies import get_current_active_user

router = APIRouter(prefix="/notifications", tags=["notifications"])


class NotificationResponse(BaseModel):
    """消息响应"""
    id: int
    type: str
    title: str
    message: str
    priority: str
    is_read: bool
    read_at: Optional[datetime] = None
    related_id: Optional[str] = None
    related_type: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class NotificationCreate(BaseModel):
    """创建消息"""
    type: str
    title: str
    message: str
    priority: str = "NORMAL"
    related_id: Optional[str] = None
    related_type: Optional[str] = None


@router.get("/", response_model=List[NotificationResponse])
async def get_notifications(
    skip: int = Query(0, ge=0),
    limit: int = Query(20, ge=1, le=100),
    is_read: Optional[bool] = None,
    type: Optional[str] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """获取用户的消息列表"""
    query = db.query(Notification).filter(Notification.user_id == current_user.id)
    
    if is_read is not None:
        query = query.filter(Notification.is_read == is_read)
    
    if type:
        query = query.filter(Notification.type == type)
    
    # 按创建时间倒序
    notifications = query.order_by(Notification.created_at.desc()).offset(skip).limit(limit).all()
    return notifications


@router.get("/unread-count", response_model=dict)
async def get_unread_count(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """获取未读消息数"""
    unread_count = db.query(Notification).filter(
        Notification.user_id == current_user.id,
        Notification.is_read == False
    ).count()
    
    return {"unread_count": unread_count}


@router.get("/{notification_id}", response_model=NotificationResponse)
async def get_notification(
    notification_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """获取单个消息"""
    notification = db.query(Notification).filter(
        Notification.id == notification_id,
        Notification.user_id == current_user.id
    ).first()
    
    if not notification:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Notification not found"
        )
    
    return notification


@router.put("/{notification_id}/read", response_model=NotificationResponse)
async def mark_as_read(
    notification_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """标记消息为已读"""
    notification = db.query(Notification).filter(
        Notification.id == notification_id,
        Notification.user_id == current_user.id
    ).first()
    
    if not notification:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Notification not found"
        )
    
    notification.mark_as_read()
    db.commit()
    db.refresh(notification)
    
    return notification


@router.put("/mark-all-as-read", response_model=dict)
async def mark_all_as_read(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """标记所有消息为已读"""
    db.query(Notification).filter(
        Notification.user_id == current_user.id,
        Notification.is_read == False
    ).update(
        {
            Notification.is_read: True,
            Notification.read_at: datetime.utcnow(),
            Notification.updated_at: datetime.utcnow()
        }
    )
    db.commit()
    
    return {"status": "ok"}


@router.delete("/{notification_id}", response_model=dict)
async def delete_notification(
    notification_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """删除消息"""
    notification = db.query(Notification).filter(
        Notification.id == notification_id,
        Notification.user_id == current_user.id
    ).first()
    
    if not notification:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Notification not found"
        )
    
    db.delete(notification)
    db.commit()
    
    return {"status": "ok"}


@router.delete("/", response_model=dict)
async def delete_all_notifications(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """删除所有消息"""
    db.query(Notification).filter(
        Notification.user_id == current_user.id
    ).delete()
    db.commit()
    
    return {"status": "ok"}

